sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala

```diff
@@ -0,0 +1,113 @@
+import org.apache.spark.sql.types.{DataType, DecimalType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+
+/**
+ * A function that converts string to numeric.
+ */
+@ExpressionDescription(
+  usage = """
+     _FUNC_(strExpr, formatExpr) - Convert `strExpr` to a number based on the `formatExpr`.
+       The format can consist of the following characters:
+         '9':  digit position (can be dropped if insignificant)
+         '0':  digit position (will not be dropped, even if insignificant)
+         '.':  decimal point (only allowed once)
+         ',':  group (thousands) separator
+         'S':  sign anchored to number (uses locale)
+         'D':  decimal point (uses locale)
+         'G':  group separator (uses locale)
+         '$':  specifies that the input value has a leading $ (Dollar) sign.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('454', '999');
+       454
+      > SELECT _FUNC_('454.00', '000D00');
+       454.00
+      > SELECT _FUNC_('12,454', '99G999');
+       12454
+      > SELECT _FUNC_('$78.12', '$99.99');
+       78.12
+      > SELECT _FUNC_('12,454.8-', '99G999D9S');
+       -12454.8
+  """,
+  since = "3.3.0",
+  group = "string_funcs")
+case class ToNumber(left: Expression, right: Expression)
+  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant {
+
+  private lazy val numberFormat = right.eval().toString.toUpperCase(Locale.ROOT)
+  private lazy val builder = new NumberFormatBuilder(numberFormat)
+  private lazy val (precision, scale) = builder.parsePrecisionAndScale()
+
+  override def dataType: DataType = DecimalType(precision, scale)
+
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val inputTypeCheck = super.checkInputDataTypes()
+    if (inputTypeCheck.isSuccess) {
+      if (right.foldable) {
+        if (builder.normalizedNumberFormat.length == 0) {
+          TypeCheckResult.TypeCheckFailure(s"Format expression cannot be empty")
+        } else {
+          builder.check()
+        }
+      } else {
+        TypeCheckResult.TypeCheckFailure(s"Format expression must be foldable, but got $right")
+      }
+    } else {
+      inputTypeCheck
+    }
+  }
+
+  override def prettyName: String = "to_number"
+
+  override def nullSafeEval(string: Any, format: Any): Any = {
+    val input = string.asInstanceOf[UTF8String]
+    builder.parse(input)
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val builderClass = classOf[NumberFormatBuilder].getName
+    val builder = ctx.freshName("builder")
+    val eval = left.genCode(ctx)
+    ev.copy(code = code"""
+      ${eval.code}
+      boolean ${ev.isNull} = ${eval.isNull};
+      ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+      if (!${ev.isNull}) {
+        $builderClass $builder = new $builderClass("$numberFormat");
```

Comment:
It's pretty risky to embed user input in the generated code, as people can inject malicious code.

We should do `ctx.addReferenceObj(builder)`